subscription.go

```diff
@@ -5,66 +5,161 @@ import (
+func (self *SubscriptableSchema) Subscribe(ctx context.Context, queryString string, operationName string, variables map[string]interface{}) (<-chan interface{}, error) {
+	c := Subscribe(Params{
+		Schema:         self.Schema,
+		Context:        ctx,
+		OperationName:  operationName,
+		RequestString:  queryString,
+		RootObject:     self.RootObject,
+		VariableValues: variables,
+	})
+	to := make(chan interface{})
+	go func() {
+		defer close(to)
+		select {
+		case <-ctx.Done():
+			return
+		case res, more := <-c:
+			if !more {
+				return
+			}
+			to <- res
+		}
+	}()
+	return to, nil
+}
+
+// Subscribe performs a subscribe operation on the given query and schema
+// currently does not support extensions hooks
+func Subscribe(p Params) chan *Result {
+
+	source := source.NewSource(&source.Source{
+		Body: []byte(p.RequestString),
+		Name: "GraphQL request",
+	})
+
+	// TODO run extensions hooks
+
+	// parse the source
+	AST, err := parser.Parse(parser.ParseParams{Source: source})
+	if err != nil {
+
+		// merge the errors from extensions and the original error from parser
+		return sendOneResultAndClose(&Result{
+			Errors: gqlerrors.FormatErrors(err),
+		})
+	}
+
+	// validate document
+	validationResult := ValidateDocument(&p.Schema, AST, nil)
+
+	if !validationResult.IsValid {
+		// run validation finish functions for extensions
+		return sendOneResultAndClose(&Result{
+			Errors: validationResult.Errors,
+		})
+
+	}
+	return ExecuteSubscription(ExecuteParams{
+		Schema:        p.Schema,
+		Root:          p.RootObject,
+		AST:           AST,
+		OperationName: p.OperationName,
+		Args:          p.VariableValues,
+		Context:       p.Context,
+	})
+}
+
+func sendOneResultAndClose(res *Result) chan *Result {
+	resultChannel := make(chan *Result, 1)
+	resultChannel <- res
+	close(resultChannel)
+	return resultChannel
+}
+
+// ExecuteSubscription is similar to graphql.Execute but returns a channel instead of a Result
+// currently does not support extensions
+func ExecuteSubscription(p ExecuteParams) chan *Result {
+
+	if p.Context == nil {
+		p.Context = context.Background()
+	}
 
 	var mapSourceToResponse = func(payload interface{}) *Result {
 		return Execute(ExecuteParams{
 			Schema:        p.Schema,
 			Root:          payload,
-			AST:           p.Document,
+			AST:           p.AST,
 			OperationName: p.OperationName,
-			Args:          p.VariableValues,
-			Context:       p.ContextValue,
+			Args:          p.Args,
+			Context:       p.Context,
 		})
 	}
-
+	var resultChannel = make(chan *Result)
 	go func() {
-		result := &Result{}
+		defer close(resultChannel)
 		defer func() {
 			if err := recover(); err != nil {
-				result.Errors = append(result.Errors, gqlerrors.FormatError(err.(error)))
-				resultChannel <- result
+				e, ok := err.(error)
+				if !ok {
+					fmt.Println("strange program path")
+					return
+				}
+				resultChannel <- &Result{
```

Comment:
if the resultChannel is closed before this runs, this will panic which is why I had it in this deferred function at the end

Reply:
the only way the channel can close is the `defer close(resultChannel)`, this means the channel will be closed after ther error handling (defer functions are run from last to first)

I will add a test case for a panic